databricks-sdk-java/src/main/java/com/databricks/sdk/core/ApiClient.java

```diff
@@ -255,37 +263,62 @@ private Response executeInner(Request in) {
       // 2. The request failed with a non-retriable error (err != null, out == null).
       // 3. The request failed with a retriable error, but the number of attempts exceeds
       // maxAttempts.
-      DatabricksError res = ApiErrors.checkForRetry(out, err);
-      if (!res.isRetriable()) {
-        if (res.getErrorCode() == null) {
+      if (!retryStrategy.isRetriable(out, err)) {
+        if (retryStrategy.getError() == null || retryStrategy.getError().getErrorCode() == null) {
```

Comment:
It feels kind of wrong to me that the retry strategy is returning the error, and it also results in internal mutability which I would like to avoid if possible.

To do this, I think we need to separate the two independent retry conditions, one based on the exception thrown by the HTTP client, the other based on the error response from Databricks.

Algorithm:
1. If exception is not null, check the exception. Retry if retriable, otherwise don't.
2. No exception. It's an error response that we can read using `ApiErrors.readErrorFromResponse`.
3. Then, check the error response for retry conditions.